// @chainsafe/discv5/lib/message/create.js
import { randomBytes, toBytes } from "@noble/hashes/utils";
import { MessageType, } from "./types.js";
export function createRequestId() {
    return randomBytes(8);
}
export function createPingMessage(enrSeq) {
    return {
        type: MessageType.PING,
        id: createRequestId(),
        enrSeq,
    };
}
export function createFindNodeMessage(distances) {
    return {
        type: MessageType.FINDNODE,
        id: createRequestId(),
        distances,
    };
}
export function createNodesMessage(id, total, enrs) {
    return {
        type: MessageType.NODES,
        id,
        total,
        enrs,
    };
}
export function createTalkRequestMessage(request, protocol) {
    return {
        type: MessageType.TALKREQ,
        id: createRequestId(),
        protocol: toBytes(protocol),
        request: toBytes(request),
    };
}
export function createTalkResponseMessage(requestId, payload) {
    return {
        type: MessageType.TALKRESP,
        id: requestId,
        response: payload,
    };
}
